framework/src/org/checkerframework/common/basetype/BaseTypeChecker.java

```diff
@@ -412,35 +425,181 @@ T getTypeFactoryOfSubchecker(Class<U> checkerClass) {
+     */
+    private void printCollectedMessages(CompilationUnitTree unit) {
+        if (messageStore != null) {
+            for (CheckerMessage pos : messageStore) {
+                super.printMessage(pos.kind, pos.message, pos.source, unit);
+            }
+        }
+    }
+
+    /**
+     * Sort by position at which the error will be printed, then by the order in which the checkers
+     * run, then by kind of message, and finally by the message string.
+     */
+    private final Comparator<CheckerMessage> checkerMessageComparator =
+            new Comparator<CheckerMessage>() {
+
+                @Override
+                public int compare(CheckerMessage o1, CheckerMessage o2) {
+                    int byPos = InternalUtils.compareDiagnosticPosition(o1.source, o2.source);
+                    if (byPos != 0) {
+                        return byPos;
+                    }
+
+                    int o1Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o1.checker);
+                    int o2Checker = BaseTypeChecker.this.getSubcheckers().indexOf(o2.checker);
+                    if (o1Checker != o2Checker) {
+                        if (o1Checker == -1) {
+                            o1Checker = BaseTypeChecker.this.getSubcheckers().size();
+                        }
+                        if (o2Checker == -1) {
+                            o2Checker = BaseTypeChecker.this.getSubcheckers().size();
+                        }
+                        return Integer.compare(o1Checker, o2Checker);
+                    }
+
+                    int kind = o1.kind.compareTo(o2.kind);
+                    if (kind != 0) {
+                        return kind;
+                    }
+
+                    return o1.message.compareTo(o2.message);
+                }
+            };
+
+    /** Represents a message (e.g., an error message) issued by a checker. */
+    private class CheckerMessage {
+        Diagnostic.Kind kind;
+        String message;
+        Tree source;
+        BaseTypeChecker checker;
+
+        private CheckerMessage(
+                Diagnostic.Kind kind, String message, Tree source, BaseTypeChecker checker) {
+            this.kind = kind;
+            this.message = message;
+            this.source = source;
+            this.checker = checker;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            CheckerMessage that = (CheckerMessage) o;
+
+            if (kind != that.kind) {
+                return false;
+            }
+            if (!message.equals(that.message)) {
+                return false;
+            }
+            if (!source.equals(that.source)) {
+                return false;
+            }
+            return checker.equals(that.checker);
```

Comment:
I think neither Tree nor Checker implement equals, so you could just use reference equality.